opm/autodiff/BlackoilAmg.hpp

```diff
@@ -0,0 +1,727 @@
+ * @tparam C1 The type of the information about the communication. Either
+ *            Dune::OwnerOverlapCopyCommunication or Dune::SequentialInformation.
+ */
+template<class O, class S, class C, class P>
+class OneStepAMGCoarseSolverPolicy
+{
+public:
+    typedef P LevelTransferPolicy;
+    /** @brief The type of the linear operator used. */
+    typedef O Operator;
+    /** @brief The type of the range and domain of the operator. */
+    typedef typename O::range_type X;
+    /** @brief The type of the crition used for the aggregation within AMG.*/
+    typedef C Criterion;
+    /** @brief The type of the communication used for AMG.*/
+    typedef typename P::ParallelInformation Communication;
+    /** @brief The type of the smoother used in AMG. */
+    typedef S Smoother;
+    /** @brief The type of the arguments used for constructing the smoother. */
+    typedef typename Dune::Amg::SmootherTraits<S>::Arguments SmootherArgs;
+    /** @brief The type of the AMG construct on the coarse level.*/
+    typedef Dune::Amg::AMG<Operator,X,Smoother,Communication> AMGType;
+    /**
+     * @brief Constructs the coarse solver policy.
+     * @param args The arguments used for constructing the smoother.
+     * @param c The crition used for the aggregation within AMG.
+     */
+    OneStepAMGCoarseSolverPolicy(const SmootherArgs& args, const Criterion& c)
+        : smootherArgs_(args), criterion_(c)
+    {}
+    /** @brief Copy constructor. */
+    OneStepAMGCoarseSolverPolicy(const OneStepAMGCoarseSolverPolicy& other)
+        : coarseOperator_(other.coarseOperator_), smootherArgs_(other.smootherArgs_),
+          criterion_(other.criterion_)
+    {}
+private:
+    /**
+     * @brief A wrapper that makes an inverse operator out of AMG.
+     *
+     * The operator will use one step of AMG to approximately solve
+     * the coarse level system.
+     */
+    struct AMGInverseOperator : public Dune::InverseOperator<X,X>
+    {
+        AMGInverseOperator(const typename AMGType::Operator& op,
+                           const Criterion& crit,
+                           const typename AMGType::SmootherArgs& args,
+                           const Communication& comm)
+            : amg_(op, crit,args, comm), op_(op), comm_(comm), first_(true)
+        {}
+
+        void apply(X& x, X& b, double reduction, Dune::InverseOperatorResult& res)
+        {
+            DUNE_UNUSED_PARAMETER(reduction);
+            DUNE_UNUSED_PARAMETER(res);
+            /*
+            if(first_)
+            {
+                amg_.pre(x,b);
+                first_=false;
+                x_=x;
+            }
+            amg_.apply(x,b);
+            */
+            using Chooser = Dune::ScalarProductChooser<X,Communication,AMGType::category>;
+            auto sp = Chooser::construct(comm_);
+            Dune::BiCGSTABSolver<X> solver(const_cast<typename AMGType::Operator&>(op_), *sp, amg_, 1e-2, 25, 0);
```

Comment:
hardcoded params, possibly forgotten?

Reply:
I guess this was fixed in a later commit (e24d338d)